onnx/onnx_backend.h

```diff
@@ -0,0 +1,914 @@
+ * backend.
+ *
+ * Value type: uint64_t.
+ * If the backend does not support FP16 computation, the value MUST be 0.
+ */
+#define ONNX_BACKEND_MACS_FP16 31
+
+/**
+ * Bandwidth, in bytes per second, of the global memory specific to the backend
+ * device.
+ *
+ * Value type: uint64_t.
+ */
+#define ONNX_BACKEND_MEMORY_BANDWIDTH 35
+
+/**
+ * Bandwidth, in bytes per second, of transferring data from cacheable
+ * CPU-allocated memory to the backend device.
+ *
+ * Value type: uint64_t.
+ */
+#define ONNX_BACKEND_CPU_MEMORY_READ_BANDWIDTH 36
+
+/**
+ * Bandwidth, in bytes per second, of transferring data to cacheable
+ * CPU-allocated memory from the backend device.
+ *
+ * Value type: uint64_t.
+ */
+#define ONNX_BACKEND_CPU_MEMORY_WRITE_BANDWIDTH 37
+
+/* Note: the data type values match ONNX TensorProto.DataType enum */
+#define ONNX_DATATYPE_UNDEFINED 0
+#define ONNX_DATATYPE_FLOAT16 10
+#define ONNX_DATATYPE_FLOAT32 1
+#define ONNX_DATATYPE_FLOAT64 11
+#define ONNX_DATATYPE_INT8 3
+#define ONNX_DATATYPE_INT16 5
+#define ONNX_DATATYPE_INT32 6
+#define ONNX_DATATYPE_INT64 7
+#define ONNX_DATATYPE_UINT8 2
+#define ONNX_DATATYPE_UINT16 4
+#define ONNX_DATATYPE_UINT32 12
+#define ONNX_DATATYPE_UINT64 13
+
+/** Cacheable CPU memory */
+#define ONNX_MEMORY_TYPE_CPU 0
+
+#define ONNX_TENSOR_DIMS_MAX 8
+
+/**
+ * Terminates the list of auxiliary backend initialization properties passed to
+ * onnxInitBackend.
+ */
+#define ONNX_BACKEND_PROPERTY_NONE 0
+
+typedef struct onnxTensorDescriptor {
+  /**
+   * Name of the blob corresponding to this tensor in the ONNX model. The name
+   * must exactly match the ValueInfoProto.name of one of the
+   * ModelProto.graph.input or ModelProto.graph.output
+   */
+  const char* name;
+  /**
+   * Data type of the elements in the tensor.
+   *
+   * Possible values:
+   *     ONNX_DATATYPE_FLOAT16
+   *     ONNX_DATATYPE_FLOAT32
+   *     ONNX_DATATYPE_INT8
+   *     ONNX_DATATYPE_INT16
+   *     ONNX_DATATYPE_INT32
+   *     ONNX_DATATYPE_UINT8
+   *     ONNX_DATATYPE_UINT16
+   *     ONNX_DATATYPE_UINT32
+   */
+  onnxEnum dataType;
+  /**
+   * Type of memory that stores the tensor.
+   *
+   * Possible values:
+   *     ONNX_MEMORY_TYPE_CPU
+   */
+  onnxEnum memoryType;
+  /**
+   * Number of dimensions in the tensor.
+   * Must be between 0 (for a scalar) and ONNX_TENSOR_DIMS_MAX.
+   */
+  uint32_t dimensions;
+  /**
+   * Dimensions of the tensor.
+   */
+  uint64_t shape[ONNX_TENSOR_DIMS_MAX];
+  /**
+   * Pointers to tensor data.
+   */
+  onnxPointer buffer;
+} onnxTensorDescriptor;
+
+/* Function pointer declarations for dynamic loading */
+typedef uint32_t(ONNX_ABI* onnxGetNumBackendsFunction)(void);
+typedef onnxStatus(ONNX_ABI* onnxGetBackendInfoFunction)(
+    uint32_t index,
+    onnxBackendInfo infoType,
+    void* infoValue,
+    size_t* infoValueSize);
+typedef onnxStatus(ONNX_ABI* onnxGetBackendCompatibilityFunction)(
+    uint32_t index,
+    size_t onnxModelSize,
+    const void* onnxModel);
```

Comment:
What's the onnxModel? It should be a sub-graph, I think.

Reply:
How do we split the whole graph into sub-graphs and dispatch to different backends please? a sub-graph partition API might be designed and provided.

Reply:
`onnxModel` is a serialized `ModelProto` message representing a subgraph. Please see above on the reasons for choosing `ModelProto` instead of `GraphProto`.

Reply:
Sub-graph partitioning is not part of the proposal: its implementation it left to high-level frameworks which use the Backend API. The Backend API, however, provides information queries to help the high-level level frameworks with this task:  `ONNX_BACKEND_MACS_FP32`/`ONNX_BACKEND_MACS_FP16` and `ONNX_BACKEND_MEMORY_BANDWIDTH` information queries get information about the target backend needed to build a roofline model of performance for any known subgraph. Combined with `ONNX_BACKEND_CPU_MEMORY_READ_BANDWIDTH` and `ONNX_BACKEND_CPU_MEMORY_WRITE_BANDWIDTH`, the model would provide an estimate for the time to offloading inputs to a subgraph to the backend, execute the subgraph on the backend, and get the results back.

Reply:
dumb question, are we expecting each backend will have full capability to run any onnx model please? so that the partitioning in high-level framework will only need to understand the resource bandwidth of each backend.

Reply:
No, backend can implement only subset of ops. That's why `onnxGetBackendCompatibility` is here - to allow hosting backend (e.g. caffe2) to ask whether particular chunk is supported or not (usually calling node by node). That's usually the approach e.g. TensorRT is integrated in TF

Reply:
Thank you for clarification! We're saying high-level callers will construct temp Model (only one node) one by one per main model to get to know the capability. Hmmm, @gramalingam is also suggesting that we should have a way to check the capability with one call without consisting of temp one-node models.